### PR TITLE
Store NoCompression entries in System.IO.Compression without wrapping them in Deflate streams, fixes #29735

### LINKED ISSUE #29735: System.IO.Compression should store NoCompression entries without wrapping them in Deflate streams
Currently when a ZipArchive is created and an entry is added with the CompressionLevel set to "no-compression" the file inside the archive is still wrapped in a deflate stream. The deflate stream correctly does not to compress the file.

However, zip does natively support uncompressed files. The compression method "stored" is made for this.

To create OpenDocument format files (ods, odt etc.) it is required to store the mimetype file without wrapping it in any compression stream. They do this, so that the mimetype can always found at a specific offset. 
This change should also be a tiny performance and archive size improvement.

I have attached a diff of the changes necessary to implement this. I've run the tests, however I've not yet been able to build corefx in a way that I can use with my app to validate the patch is working.
[diff.zip](https://github.com/dotnet/corefx/files/2008533/diff.zip)

If this is something you think is possible to merge, I can spend more time on this to create a proper PR.

### PR DESCRIPTION
**Pull request by Marv51**
This implements the feature enhancement described in #29735.

If a ZipArchiveEntry is created with CompressionLevel.NoCompression, with this change the content is no longer wrapped in a useless deflate stream. Instead it is stored directly in the archive with CompressionMethod "stored".
I added a test, that validates that the fix is working.

Please feel free to change make changes to the code, I'm sure it can be improved in a lot of ways.

### CODE REVIEW

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 1]
```diff
@@ -594,15 +602,21 @@ private CheckSumAndSizeWriteStream GetDataCompressor(Stream backingStream, bool
...
-            // call through this function for that - we just write the stored value in the header
-            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate);
-
-            Stream compressorStream = _compressionLevel.HasValue ?
-                new DeflateStream(backingStream, _compressionLevel.Value, leaveBackingStreamOpen) :
-                new DeflateStream(backingStream, CompressionMode.Compress, leaveBackingStreamOpen);
-
             bool isIntermediateStream = true;
+            Stream compressorStream;
+            if (!_compressionLevel.HasValue)
```
Comment (danmosemsft) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: I would first set the value to Compress if it has no value. Then test the value, using a switch statement.

File: src/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs [hunk 2]
```diff
@@ -115,6 +115,48 @@ public static async Task CreateNormal_Unicode_Unseekable()
...
+                    var newEntry = zip.CreateEntry(testfilename, CompressionLevel.NoCompression);
+                    using (var writer = new StreamWriter(newEntry.Open(), utf8WithoutBom))
+                    {
+                        writer.Write(testFileContent);
+                        writer.Flush();
+                    }
+                    var fileContent = s.ToArray();
+                    // zip file header stores values as little-endian
+                    var compressionMethod = fileContent[8];
+                    Assert.Equal(0, compressionMethod); // stored => 0, deflate => 8
```
Comment (ahsonkhan) [inline src/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs]: This assert is failing:
https://mc.dot.net/#/user/Marv51/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/1ccf010a41e3ad7031e1339e309cd2d95f830c3e/workItem/System.IO.Compression.Tests/analysis/xunit/System.IO.Compression.Tests.zip_CreateTests~2FCreateUncompressedArchive

```text
Message :
Assert.Equal() Failure
Expected: 0
Actual:   8
Stack Trace :
   at System.IO.Compression.Tests.zip_CreateTests.CreateUncompressedArchive() in D:\j\workspace\windows-TGrou---2a8f9c29\src\System.IO.Compression\tests\ZipArchive\zip_CreateTests.cs:line 139
```

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 3]
```diff
@@ -91,6 +91,14 @@ internal ZipArchiveEntry(ZipArchive archive, string entryName, CompressionLevel
             : this(archive, entryName)
         {
             _compressionLevel = compressionLevel;
+            if (_compressionLevel == CompressionLevel.NoCompression)
+            {
+                CompressionMethod = CompressionMethodValues.Stored;
+            }
+            else
+            {
+                CompressionMethod = CompressionMethodValues.Deflate;
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: `_compressionLevel` is Nullable, which means if someone passes in null, we set CompressionMode to Deflate, which we may not want.

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 4]
```diff
@@ -593,16 +597,23 @@ internal void ThrowIfNotOpenable(bool needToUncompress, bool needToLoadIntoMemor
...
-            // call through this function for that - we just write the stored value in the header
-            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate);
-
-            Stream compressorStream = _compressionLevel.HasValue ?
-                new DeflateStream(backingStream, _compressionLevel.Value, leaveBackingStreamOpen) :
-                new DeflateStream(backingStream, CompressionMode.Compress, leaveBackingStreamOpen);
+            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate || CompressionMethod == CompressionMethodValues.Stored);
 
             bool isIntermediateStream = true;
+            Stream compressorStream;
```
Comment (ahsonkhan) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]:
```text
// we should always be compressing with deflate. Stored is used for empty files, but we don't actually	
// call through this function for that - we just write the stored value in the header
```


Can we keep and update the relevant pieces of the above comment that got removed?

**Something like:**
```text
If `CompressionMethod` is set to something other than `Stored`, we should always be compressing with deflate.
However, if `_compressionLevel` is `NoCompression`, do not use `DeflateStream`.
```

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 5]
```diff
@@ -593,16 +597,23 @@ internal void ThrowIfNotOpenable(bool needToUncompress, bool needToLoadIntoMemor
...
+            if (!_compressionLevel.HasValue)
+            {
+                _compressionLevel = CompressionMode.Compress;
```
Comment (ahsonkhan) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: `CompressionLeve`l != `CompressionMode`

Can't assign one to another. I am assuming this should be `CompressionLevel.NoCompression`?

Comment (Marv51) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Oh I'm sorry it auto updated the pull request. I'm currently adding a switch statement there

Comment (Marv51) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: OK, I think with the 25e3f1f commit I should have addressed most of your feedback. The test does pass locally but the CI test is probably still failing. I run `..\..\build .` from inside the `\src\System.IO.Compression` directory.

File: src/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs [hunk 6]
```diff
@@ -115,6 +115,44 @@ public static async Task CreateNormal_Unicode_Unseekable()
...
+                    Assert.Equal(testfilename, readFileName);
+                    string readFileContent = ReadStringFromSpan(fileContent.AsSpan(30 + filenamelength, testFileContent.Length));
+                    Assert.Equal(testFileContent, readFileContent);
+                }
+            }
+        }
+
+        private static string ReadStringFromSpan(Span<byte> input)
+        {
+            return Text.Encoding.UTF8.GetString(input);
```
Comment (ahsonkhan) [inline src/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs]: https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netfx+CGroup_Release+AGroup_x86+TestOuter_false_prtest/18111/console
```text
00:16:43 ZipArchive\zip_CreateTests.cs(154,49): error CS1503: Argument 1: cannot convert from 'System.Span<byte>' to 'byte[]'
```

This failure is coming from the fact that we don't have the span-based `Encoding.UTF8.GetString` overload on netfx. So, it's trying to match it to the byte[] overload. Sorry about that. I didn't take netfx into account.

For simplicity, I would just call ToArray on the span:

```C#
return Text.Encoding.UTF8.GetString(input.ToArray());
```

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 7]
```diff
@@ -594,16 +598,28 @@ private CheckSumAndSizeWriteStream GetDataCompressor(Stream backingStream, bool
...
-            // call through this function for that - we just write the stored value in the header
-            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate);
-
-            Stream compressorStream = _compressionLevel.HasValue ?
-                new DeflateStream(backingStream, _compressionLevel.Value, leaveBackingStreamOpen) :
-                new DeflateStream(backingStream, CompressionMode.Compress, leaveBackingStreamOpen);
+            // By default we compress with deflate, except if compression level is set to NoCompression then stored is used.
+            // Stored is also used for empty files, but we don't actually call through this function for that - we just write the stored value in the header
+            // Deflate64 is not supported on all platforms
+            Debug.Assert(CompressionMethod == CompressionMethodValues.Deflate
```
[the remark this answers is not in the document]
Comment (Marv51) [reply src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Hmm, I looked that. I didn't change anything with regard to Deflate64. However, I don't understand how Deflate64 is not valid here, it is across most of the class. Will remove this anyway so it isn't an issue.

**Comment by ahsonkhan** [comment]
@dotnet-bot test NETFX x86 Release Build

@Marv51, can you reproduce this failure locally, for netfx?
https://mc.dot.net/#/user/Marv51/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/c626634aea7e2d3e973b932a0e68d0f5a64e4057/workItem/System.IO.Compression.Tests/analysis/xunit/System.IO.Compression.Tests.zip_CreateTests~2FCreateUncompressedArchive

Marv51  pr/jenkins/dotnet/corefx/master/  test/functional/cli/  User Build  c626634aea7e2d3e973b932a0e68d0f5a64e4057  System.IO.Compression.Tests  xunit  System.IO.Compression.Tests.zip_CreateTests/CreateUncompressedArchive
Windows.10.Amd64.ClientRS4.Open-x86-Release
```text
Message :
Assert.Equal() Failure
Expected: 0
Actual:   8
Stack Trace :
   at System.IO.Compression.Tests.zip_CreateTests.CreateUncompressedArchive() in D:\j\workspace\windows-TGrou---2a8f9c29\src\System.IO.Compression\tests\ZipArchive\zip_CreateTests.cs:line 139
```

How to build/test netfx:
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md
- Repo root: `build -framework netfx -release`
- Navigate to your System.IO.Compression.Tests folder: `dotnet msbuild /p:TargetGroup=netfx /p:ConfigurationGroup=Release`

**Comment by Marv51** [comment]
@dotnet-bot test Windows x86 Release Build

**Comment by ahsonkhan** [comment]
> can you take another look at this and maybe help me with that failing netfx test?

OK, I just realized what the issue is.
We only build this library for netcoreapp and uap (and not for netfx/netstandard) since https://github.com/dotnet/corefx/pull/16045 / https://github.com/dotnet/corefx/pull/23611.

However, we are still building tests for netstandard:
https://github.com/dotnet/corefx/blob/266ba84eae00fcafa114efe7bdb397354e69dc06/src/System.IO.Compression/tests/Configurations.props#L6

That's why the tests are failing on netfx. Your library changes aren't getting built for it so the tests are running against an older version which doesn't have your change.

**Here is the fix:**

Since we are no longer building for netstandard, I would remove the following BuildConfigurations from tests\Configurations.props:
- netstandard-Unix;
- netstandard-Windows_NT;

And update the configurations in the test csproj (remove the netstandard-* configs):
https://github.com/dotnet/corefx/blob/266ba84eae00fcafa114efe7bdb397354e69dc06/src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj#L5

cc @ianhays, @ericstj

File: src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj [hunk 8]
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <ProjectGuid>{17DA7FB5-4370-4385-9A02-FFEF9F482903}</ProjectGuid>
     <DefineConstants Condition="'$(TargetGroup)' == 'netcoreapp' or '$(TargetGroup)' == 'uap'">$(DefineConstants);netcoreapp;STREAM_MEMORY_OVERLOADS_AVAILABLE</DefineConstants>
```
Comment (ViktorHofer) [inline src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj]: You can remove the DefineConstants here and all `#if netcoreapp` and `#if STREAM_MEMORY_OVERLOADS_AVAILABLE` now as we don't compile against netstandard anymore.

Comment (Marv51) [reply src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj]: I haven't found any usage of `netcoreapp`

`STREAM_MEMORY_OVERLOADS_AVAILABLE` is used in `Common/tests/../CompressionStreamunitTestBase.cs` which is also included from System.IO.Compression.Brotli.Tests. Should I also remove netfx from that project, or should we keep this limited in scope?

Comment (ViktorHofer) [reply src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj]: System.IO.Compression.Brotli.Tests doesn't have a netfx configuration? It seems someone forgot to clean that project up. You can remove the DefineConstants there too and please remove the uap configuration in `Configurations` there as it seems to be a leftover.

File: src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj [hunk 9]
```diff
@@ -13,6 +12,9 @@
     <Compile Include="ZipArchive\zip_UpdateTests.cs" />
     <Compile Include="Utilities\StripHeaderAndFooter.cs" />
     <Compile Include="Utilities\WrappedStream.cs" />
+    <Compile Include="ZipArchive\zip_netcoreappTests.cs" />
+    <Compile Include="CompressionStreamUnitTests.Deflate.cs" />
```
Comment (ViktorHofer) [inline src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj]: nit: this and the following include should be at the top of the list. We usually sort the alphabetically.

File: src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj (hunk 8 above)
Comment (ahsonkhan) [reply src/System.IO.Compression/tests/System.IO.Compression.Tests.csproj]: Maybe this clean up should happen as a separate PR, to unblock the scoped fix for #29735.

**Comment by ahsonkhan** [comment]
@Marv51, any updates on this PR?

**Comment by Marv51** [comment]
> @Marv51, any updates on this PR?

Thanks for checking in, I've been very busy this last week. I can work on this tomorrow. I think I only need to clean up/fix some of the changes I've made to the project files and build constants.

**Comment by Marv51** [comment]
I have reverted all changes outside of System.IO.Compression and hope all tests pass now (again). 
I would like create a separate PR to clean up common tests and Brotli compression project files. Should I create a separate issue for that?

**Comment by ahsonkhan** [comment]
> I would like create a separate PR to clean up common tests and Brotli compression project files. Should I create a separate issue for that?

Yes please.

**Comment by Marv51** [comment]
Please let me know if there is anything else I should be doing to move this along. (cc: @ahsonkhan)

**Comment by ahsonkhan** [comment]
Nope, it all looks good to me. Tyvm!

**Pull request merged**